### PR TITLE
fix: call setState before callbacks

### PR DESCRIPTION
**Pull request by leogiertz**
**What kind of change does this PR introduce?**

- [x] bugfix
- [ ] feature
- [ ] refactoring / style
- [ ] build / chore
- [ ] documentation

**Did you add tests for your changes?**

- [ ] Yes, my code is well tested
- [x] Not relevant

**If relevant, did you update the documentation?**

- [ ] Yes, I've updated the documentation
- [x] Not relevant

**Summary**

A potential fix for: https://github.com/react-dropzone/react-dropzone/issues/689

**Does this PR introduce a breaking change?**

Nope

**Other information**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -221,6 +221,11 @@ class Dropzone extends React.Component {
         rejectedFiles.push(...acceptedFiles.splice(0))
       }
 
+      // Update `acceptedFiles` and `rejectedFiles` state
+      // This will make children render functions receive the appropriate
+      // values
+      this.setState({ acceptedFiles, rejectedFiles })
```
Comment (rolandjitsu) [inline src/index.js]: Actually, this still does not have any guarantee that the cb fn will be called after the state is set (`.setState()` is async).

What you want is to have the onDrop* cb fns called in the `.setState({...}, () => /* call the onDrop* fns here */)` cb fn 😄

Also, add a test for this.

File: src/index.js [hunk 2]
```diff
@@ -221,22 +221,22 @@ class Dropzone extends React.Component {
...
-
-      if (acceptedFiles.length > 0 && onDropAccepted) {
-        onDropAccepted.call(this, acceptedFiles, evt)
-      }
-
       // Update `acceptedFiles` and `rejectedFiles` state
       // This will make children render functions receive the appropriate
       // values
-      this.setState({ acceptedFiles, rejectedFiles })
+      this.setState({ acceptedFiles, rejectedFiles }, () => {
```
[the remark this answers is not in the document]
Comment (rolandjitsu) [reply src/index.js]: Hmm, that is a good question. You need to replicate the behaviour you explained:

- Mount the component
- Add a spy on the `.setState()` method
- Add the `{onDrop}` prop with your cb
- Simulate a `drop` evt
- Unmount the component inside your cb (using enzyme [.unmount()](https://airbnb.io/enzyme/docs/api/ReactWrapper/unmount.html) on the mounted component)
- Check if `.setState()` was called after your cb

Something along those lines.

[the remark this answers is not in the document]
Comment (leogiertz) [reply src/index.js]: @rolandjitsu It was a bit trickier than I expected and the test doesn't cover all of the different possible scenarios, but it checks that we're not calling setState after the onDrop callback. Thanks for the suggestion!

File: src/index.spec.js [hunk 3]
```diff
@@ -684,6 +684,21 @@ describe('Dropzone', () => {
       expect(dropzone.find(DummyChildComponent)).toHaveProp('rejectedFiles', [])
     })
 
+    it('should not set state after callbacks', async () => {
```
Comment (rolandjitsu) [inline src/index.spec.js]: Move this to the `'behavior'` test suite.

**Comment by leogiertz** [comment]
Ugh, wrong button.

**Pull request merged**